cvat/apps/engine/task.py

```diff
@@ -61,11 +62,37 @@ class SegmentsParams(NamedTuple):
     segment_size: int
     overlap: int
 
-def _copy_data_from_source(server_files, upload_dir, server_dir=None):
+def _copy_data_from_share_point(
+    server_files: List[str],
+    upload_dir: str,
+    server_dir: Optional[str] = None,
+    data_to_be_excluded: Optional[List[str]] = None,
+):
     job = rq.get_current_job()
     job.meta['status'] = 'Data are being copied from source..'
     job.save_meta()
 
+    # filter data from files/directories that should be excluded
+    if data_to_be_excluded:
+
+        expanded_server_files = []
+
+        for f in server_files:
+            path = Path(server_dir or settings.SHARE_ROOT) / Path(f)
+            if not path.is_dir():
+                expanded_server_files.append(f)
+            else:
+                for root, _, files in os.walk(str(path)):
+                    if files:
+                        expanded_server_files.extend([str(Path(f) / Path(root).relative_to(path) / Path(i)) for i in files])
+
+        server_files = expanded_server_files
+
+        server_files = list(filter(
+            lambda x: x not in data_to_be_excluded and f'{Path(x).parent}/' not in data_to_be_excluded,
```

Comment:
@SpecLad 

> Moreover, x is a path in the local syntax, whereas data_to_be_excluded always contains Unix paths.

Actually, the problem exists not only for ``data_to_be_excluded``, but probably also for ``server_files``. And for browsing shared storage, because client always uses and returns ''/'' as a separator. 

Since it is not a new in this patch, and looks like requiring significant efforts to resolve, I would suggest to create an issue and consider it as a different tasks. BTW, I personally don't remember someone claims about it.

Reply:
> I'm not sure this is correct, as x is (I think?) an absolute path, while data_to_be_excluded contains paths relative to the share root. Moreover, x is a path in the local syntax, whereas data_to_be_excluded always contains Unix paths.

No, x is a relative path.

> It's not sufficient to only check Path(x).parent, since a path might be excluded by its grandparent (or great-grandparent, etc...).

Fixed

> I don't think we need to require excluded items that are directories to end with a slash. Is there any benefit from doing so?

Yes, in this case, they are easier to handle and differentiate.

Reply:
> Yes, in this case, they are easier to handle and differentiate.

Are they, though? All I can see is that you have to append an extra slash every time you want to check if a directory is in `server_files_exclude`.

Reply:
I guess that it will help on the client side when the list with all files inside the directory will be unchecked and added to `server_files_exclude` as dir. The one approach should be used for `server_files` and `server_files_exclude`.
And the directory in the server_files should be ended with a slash because share directory can be handled without slash but cloud storage "directory" without slash is just a prefix, not directory.

Reply:
Okay, that makes sense.